File: implementations/python/pr_plan/git_analysis.py
from __future__ import annotations

import os
import re
import subprocess
from dataclasses import dataclass, replace
from pathlib import Path

from pr_plan.model import AnalysisInput, ConflictEdge, PrNumber, PullRequest


class GitError(RuntimeError):
    """A Git invocation or repository check failed."""


@dataclass(frozen=True, slots=True)
class CommandResult:
    returncode: int
    stdout: bytes
    stderr: bytes


class GitRepository:
    def __init__(self, path: Path, *, timeout_seconds: float = 30.0) -> None:
        try:
            resolved = path.resolve(strict=True)
        except OSError as error:
            raise GitError(f"{path}: {error.strerror or error}") from error
        if not resolved.is_dir():
            raise GitError(f"{path}: Git directory must be a directory")
        self.path = resolved
        self.timeout_seconds = timeout_seconds
        self.run(("rev-parse", "--git-dir"))

    def run(
        self, args: tuple[str, ...], *, expected: frozenset[int] = frozenset({0})
    ) -> CommandResult:
        command = ("git", "-C", os.fspath(self.path), *args)
        environment = os.environ.copy()
        for variable in (
            "GIT_ALTERNATE_OBJECT_DIRECTORIES",
            "GIT_COMMON_DIR",
            "GIT_CONFIG_COUNT",
            "GIT_CONFIG_PARAMETERS",
            "GIT_DIR",
            "GIT_INDEX_FILE",
            "GIT_OBJECT_DIRECTORY",
            "GIT_WORK_TREE",
        ):
            environment.pop(variable, None)
        environment.update(
            {
                "GIT_CONFIG_NOSYSTEM": "1",
                "GIT_CONFIG_GLOBAL": os.devnull,
                "GIT_OPTIONAL_LOCKS": "0",
                "GIT_TERMINAL_PROMPT": "0",
                "LC_ALL": "C",
            }
        )
        try:
            completed: subprocess.CompletedProcess[bytes] = subprocess.run(
                command,
                check=False,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                env=environment,
                timeout=self.timeout_seconds,
            )
        except FileNotFoundError as error:
            raise GitError("git executable was not found") from error
        except subprocess.TimeoutExpired as error:
            raise GitError(
                f"git {args[0]} timed out after {self.timeout_seconds:g} seconds"
            ) from error
        result = CommandResult(completed.returncode, completed.stdout, completed.stderr)
        if result.returncode not in expected:
            detail = result.stderr.decode("utf-8", errors="replace").strip()
            suffix = f": {detail}" if detail else ""
            raise GitError(
                f"git {args[0]} exited with status {result.returncode}{suffix}"
            )
        return result

    def resolve_commit(self, revision: str) -> str:
        result = self.run(("rev-parse", "--verify", f"{revision}^{{commit}}"))
        object_id = result.stdout.decode("ascii", errors="strict").strip()
        if re.fullmatch(r"[0-9a-fA-F]{40}|[0-9a-fA-F]{64}", object_id) is None:
            raise GitError("git rev-parse returned an invalid commit ID")
        return object_id.lower()

    def merge_base(self, left: str, right: str) -> str:
        result = self.run(("merge-base", left, right))
        object_id = result.stdout.decode("ascii", errors="strict").strip()
        if re.fullmatch(r"[0-9a-fA-F]{40}|[0-9a-fA-F]{64}", object_id) is None:
            raise GitError("git merge-base returned an invalid commit ID")
        return object_id.lower()

    def changed_files(self, base: str, head: str) -> tuple[str, ...]:
        common = self.merge_base(base, head)
        result = self.run(("diff", "--name-only", "-z", common, head, "--"))
        return _nul_paths(result.stdout)

    def conflict_paths(self, left: str, right: str) -> tuple[str, ...]:
        result = self.run(
            (
                "merge-tree",
                "--write-tree",
                "--name-only",
                "--no-messages",
                "-z",
                left,
                right,
            ),
            expected=frozenset({0, 1}),
        )
        if result.returncode == 0:
            return ()
        records = result.stdout.split(b"\0")
        # In --write-tree mode the first record is always the resulting tree ID.
        return _path_records(records[1:])

    def is_ancestor(self, before: str, after: str) -> bool:
        result = self.run(
            ("merge-base", "--is-ancestor", before, after),
            expected=frozenset({0, 1}),
        )
        return result.returncode == 0


def _path_records(records: list[bytes]) -> tuple[str, ...]:
    return tuple(sorted({os.fsdecode(record) for record in records if record}))


def _nul_paths(output: bytes) -> tuple[str, ...]:
    return _path_records(output.split(b"\0"))


def analyze_repository(data: AnalysisInput, path: Path) -> AnalysisInput:
    repository = GitRepository(path)
    resolved_heads: dict[PrNumber, str] = {}
    analyzed_prs: list[PullRequest] = []
    for pr in data.prs:
        if pr.git_head is None or pr.git_base is None:
            raise GitError(f"internal error: missing Git revisions for PR #{pr.number}")
        head = repository.resolve_commit(pr.git_head)
        base = repository.resolve_commit(pr.git_base)
        resolved_heads[pr.number] = head
        analyzed_prs.append(
            replace(
                pr,
                files=repository.changed_files(base, head),
                base_conflict_paths=repository.conflict_paths(base, head),
            )
        )

    conflicts: list[ConflictEdge] = []
    ancestry: list[tuple[PrNumber, PrNumber]] = []
    for left_index, left in enumerate(data.prs):
        left_head = resolved_heads[left.number]
        for right in data.prs[left_index + 1 :]:
            right_head = resolved_heads[right.number]
            paths = repository.conflict_paths(left_head, right_head)
            if paths:
                conflicts.append(ConflictEdge(left.number, right.number, paths))
            if repository.is_ancestor(left_head, right_head):
                ancestry.append((left.number, right.number))
            if repository.is_ancestor(right_head, left_head):
                ancestry.append((right.number, left.number))

    return AnalysisInput(
        repository=data.repository,
        prs=tuple(analyzed_prs),
        conflict_edges=tuple(conflicts),
        ancestry_edges=tuple(sorted(ancestry)),
    )

File: implementations/nim/git_analysis.nim
import std/[algorithm, monotimes, os, osproc, posix, sets, strtabs,
  strutils, times, unicode, tables]

import model

type
  GitError* = object of CatchableError

  CommandResult = object
    status: int
    stdout: string
    stderr: string

  GitRepository = object
    path: string
    executable: string

const blockedGitEnvironment = [
  "GIT_ALTERNATE_OBJECT_DIRECTORIES", "GIT_COMMON_DIR", "GIT_CONFIG_COUNT",
  "GIT_CONFIG_PARAMETERS", "GIT_DIR", "GIT_EDITOR", "GIT_EXEC_PATH",
  "GIT_EXTERNAL_DIFF", "GIT_INDEX_FILE", "GIT_OBJECT_DIRECTORY", "GIT_PAGER",
  "GIT_SEQUENCE_EDITOR", "GIT_SSH", "GIT_SSH_COMMAND", "GIT_WORK_TREE"
]

proc gitError(message: string): ref GitError =
  newException(GitError, message)

proc gitEnvironment(): StringTableRef =
  result = newStringTable(modeCaseSensitive)
  let blocked = blockedGitEnvironment.toHashSet()
  for name, value in envPairs():
    if name notin blocked:
      result[name] = value
  result["GIT_CONFIG_NOSYSTEM"] = "1"
  result["GIT_CONFIG_GLOBAL"] = "/dev/null"
  result["GIT_CONFIG_COUNT"] = "0"
  result["GIT_OPTIONAL_LOCKS"] = "0"
  result["GIT_TERMINAL_PROMPT"] = "0"
  result["LC_ALL"] = "C"

proc readAvailable(handle: FileHandle, destination: var string): bool =
  var buffer: array[8192, char]
  while true:
    let count = posix.read(cint(handle), addr buffer[0], buffer.len)
    if count > 0:
      let previousLength = destination.len
      destination.setLen(previousLength + int(count))
      copyMem(addr destination[previousLength], addr buffer[0], int(count))
      return true
    if count == 0:
      return false
    if errno == EINTR:
      continue
    if errno == EAGAIN or errno == EWOULDBLOCK:
      return true
    raiseOSError(osLastError())

proc capture(command: string, arguments: seq[string],
    environment: StringTableRef): CommandResult =
  var process: Process
  try:
    process = startProcess(command, args = arguments, env = environment,
      options = {})
  except OSError, IOError:
    raise gitError("could not run git: " & getCurrentExceptionMsg())
  defer: process.close()

  var stdoutOpen = true
  var stderrOpen = true
  let deadline = getMonoTime() + initDuration(seconds = 30)
  while stdoutOpen or stderrOpen:
    if getMonoTime() >= deadline:
      process.terminate()
      discard process.waitForExit(500)
      raise gitError("git command timed out after 30 seconds")
    var descriptors = [
      TPollfd(fd: if stdoutOpen: cint(process.outputHandle()) else: -1,
        events: POLLIN),
      TPollfd(fd: if stderrOpen: cint(process.errorHandle()) else: -1,
        events: POLLIN)
    ]
    let ready = posix.poll(addr descriptors[0], Tnfds(descriptors.len), 100)
    if ready < 0:
      if errno == EINTR:
        continue
      raiseOSError(osLastError())
    if stdoutOpen and
        (descriptors[0].revents and (POLLIN or POLLHUP or POLLERR)) != 0:
      stdoutOpen = readAvailable(process.outputHandle(), result.stdout)
    if stderrOpen and
        (descriptors[1].revents and (POLLIN or POLLHUP or POLLERR)) != 0:
      stderrOpen = readAvailable(process.errorHandle(), result.stderr)
  result.status = process.waitForExit()

proc run(repository: GitRepository, arguments: seq[string],
    allowed: openArray[int] = [0]): CommandResult =
  if arguments.len == 0:
    raise gitError("internal error: empty Git command")
  result = capture(repository.executable, @["-C", repository.path] & arguments,
    gitEnvironment())
  if result.status notin allowed:
    var detail = result.stderr.strip()
    if detail.len == 0:
      detail = result.stdout.strip()
    let suffix = if detail.len == 0: "" else: ": " & detail
    raise gitError("git " & arguments[0] & " exited with status " &
      $result.status & suffix)

proc openGitRepository(path: string): GitRepository =
  try:
    result.path = expandFilename(path)
  except OSError:
    raise gitError(path & ": " & getCurrentExceptionMsg())
  if not dirExists(result.path):
    raise gitError(path & ": Git directory must be a directory")
  result.executable = findExe("git")
  if result.executable.len == 0:
    raise gitError("could not find git on PATH")
  discard result.run(@["rev-parse", "--git-dir"])

proc validObjectId(value: string): bool =
  if value.len notin [40, 64]:
    return false
  for character in value:
    if character notin {'0'..'9', 'a'..'f', 'A'..'F'}:
      return false
  true

proc resolveCommit(repository: GitRepository, revision: string): string =
  let command = repository.run(@["rev-parse", "--verify", revision & "^{commit}"])
  result = command.stdout.strip().toLowerAscii()
  if not validObjectId(result):
    raise gitError("git rev-parse returned an invalid commit ID")

proc mergeBase(repository: GitRepository, left, right: string): string =
  let command = repository.run(@["merge-base", left, right])
  result = command.stdout.strip().toLowerAscii()
  if not validObjectId(result):
    raise gitError("git merge-base returned an invalid commit ID")

proc nulPaths(output: string): seq[string] =
  var seen = initHashSet[string]()
  for path in output.split('\0'):
    if path.len == 0:
      continue
    if validateUtf8(path) != -1:
      raise gitError("git returned a path that is not valid UTF-8")
    if path notin seen:
      seen.incl(path)
      result.add(path)
  result.sort()

proc changedFiles(repository: GitRepository, base, head: string): seq[string] =
  let common = repository.mergeBase(base, head)
  let command = repository.run(@[
    "diff", "--no-ext-diff", "--name-only", "-z", common, head, "--"])
  nulPaths(command.stdout)

proc conflictPaths(repository: GitRepository, left, right: string): seq[string] =
  let command = repository.run(@[
    "merge-tree", "--write-tree", "--name-only", "--no-messages", "-z",
    left, right], [0, 1])
  if command.status == 0:
    return @[]
  let records = command.stdout.split('\0')
  if records.len < 2 or not validObjectId(records[0]):
    raise gitError("git merge-tree returned malformed conflict output")
  nulPaths(records[1 .. ^1].join("\0"))

proc isAncestor(repository: GitRepository, before, after: string): bool =
  repository.run(@["merge-base", "--is-ancestor", before, after], [0, 1]).status == 0

proc analyzeRepository*(data: AnalysisInput, path: string): AnalysisInput =
  let repository = openGitRepository(path)
  var resolvedHeads = initTable[PrNumber, string]()
  result.repository = data.repository
  for inputPr in data.prs:
    var pr = inputPr
    if pr.gitHead.len == 0 or pr.gitBase.len == 0:
      raise gitError("internal error: missing Git revisions for PR #" & $pr.number)
    let head = repository.resolveCommit(pr.gitHead)
    let base = repository.resolveCommit(pr.gitBase)
    pr.files = repository.changedFiles(base, head)
    pr.baseConflictPaths = repository.conflictPaths(base, head)
    resolvedHeads[pr.number] = head
    result.prs.add(pr)

  for leftIndex in 0 ..< result.prs.len:
    let left = result.prs[leftIndex]
    let leftHead = resolvedHeads[left.number]
    for rightIndex in leftIndex + 1 ..< result.prs.len:
      let right = result.prs[rightIndex]
      let rightHead = resolvedHeads[right.number]
      let paths = repository.conflictPaths(leftHead, rightHead)
      if paths.len > 0:
        result.conflictEdges.add(PathEdge(
          a: left.number, b: right.number, paths: paths))
      if leftHead == rightHead:
        continue
      if repository.isAncestor(leftHead, rightHead):
        result.ancestryEdges.add(AncestryEdge(
          before: left.number, after: right.number))
      elif repository.isAncestor(rightHead, leftHead):
        result.ancestryEdges.add(AncestryEdge(
          before: right.number, after: left.number))
